x-pack/plugins/security/server/authentication/providers/basic.ts

```diff
@@ -31,6 +31,16 @@ interface ProviderState {
   authorization?: string;
 }
 
+/**
+ * Checks whether current request can initiate new session.
+ * @param request Request instance.
+ */
+function canStartNewSession(request: KibanaRequest) {
```

Comment:
__note:__ this new method isn't required for basic/saml/oidc/token providers since only `canRedirectRequest(request)` would be enough (these providers return `AuthenticationResult.redirected()` and Core treats it as `notHandled` if route has __optional__ authentication). I added this new method so that all providers are consistent with pki and kerberos providers where it's necessary. 

The idea here is based on the assumption that `optional` authentication mode is the mode that allows request processing flow to pick up existing user information if it's available (e.g. when request is attributed with the existing session cookie or authorization headers). 

Even though such request can cause token refresh (and extend existing session as the result) it's not supposed to create a brand new session. This requirement is implicitly enforced by basic/token/saml/oidc providers since they need user involvement to perform authentication and create a new session. But Kerberos and PKI are different, they don't redirect user anywhere and use different mechanisms to establish new session (SPNEGO or peer certificate validation) and hence need to rely on  `request.route.options.authRequired` to know if it's allowed to create a new session or not.

The main reason for this change is that I wanted to use `optional` authentication mode for the Login Selector view route and support PKI/Kerberos as login methods there. And we don't want to automatically create new PKI/Kerberos session when we visit Login Selector view. Moreover we have `/api/capabilities/` endpoint that is called for every view (whether it requires authentication or not) and this endpoint uses `optional` authentication mode.

What do you think? Am I missing something?

Reply:
>  and Core treats it as notHandled if route has optional authentication). 

It looks like we should consolidate all this logic in the Security plugin. The problem with the current implementation: - it's scattered different layers it's really hard to grasp the auth flow.
- core and security have to align `redirect` logic

Reply:
> It looks like we should consolidate all this logic in the Security plugin. The problem with the current implementation:

Yep, definitely agree, that's why I explicitly opted out of relying on `redirects` in Security code for now when `optional` auth mode is used. I want to battle test current implementation and if we don't notice any other issues we'd need consolidate this logic somehow.